docs/kuneiform/procedures.md

```diff
@@ -0,0 +1,275 @@
+---
+sidebar_position: 4
+sidebar_label: "Procedures"
+id: kuneiform-procedures
+title: Kuneiform Procedures
+description: An outline of Kuneiform procedures.
+slug: /kuneiform/procedures
+---
+
+Kuneiform supports a basic procedural language for building logic into schemas. Procedures can be used for building access control logic, performing arithmetic, and managing control flow.
+
+## Procedure Blocks
+
+Procedures are declared using the `procedure` key word. They can be given a list of parameters and their corresponding types, return types, and access modifiers. The example below shows the signature for a `get_user` procedure. It takes one input (the id of the user, of type `uuid`), and returns the user's name and age. It will return exactly one record, containing a name and age:
+
+```ts
+procedure get_user($id uuid) public view returns (name text, age int) {
+    // procedure implementation
+}
+```
+
+Procedures can also return tables, which are a collection of many records. The below example returns all user IDs and names of a certain age:
+
+```ts
+procedure get_users($age int) public view returns table(id uuid, name text) {
+    // procedure implementation
+}
+```
+
+If a procedure does not return anything, the return should not be specified:
+
+```ts
+procedure create_user($name text, $age int) public {
+    // procedure implementation
+}
+```
+
+## Procedure Syntax
+
+### Declaring and Assigning Variables
+
+All variables in procedures are strongly typed. They can be assigned at declaration, or assigned to after declaration. They can also be type casted:
+
+```ts
+procedure declare_vars() public {
+    // declare and assign new int var:
+    $var1 int := 1;
+
+    // declare a new var to be assigned to later
+    $var2 text;
+
+    // assign to $var2. type cast to match type text:
+    $var2 := $var1::text;
```

Comment:
A question that are reader is likely to have is, what type casts are valid?  For instance, what if someone did `$var3 text := 'nan'` and then `$var3::int`?

Reply:
I sort've understand what you're saying, but I am not sure how I would document this, without documenting every possible typecast